vsp.go

```diff
@@ -60,62 +66,298 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+
+				// Skip this purchase when no more tickets may be purchased in the interval and
+				// the next sdiff is unknown.  The earliest any ticket may be mined is two
+				// blocks from now, with the next block containing the split transaction
+				// that the ticket purchase spends.
+				if height+2 == nextIntervalStart {
+					log.Debugf("Skipping purchase: next sdiff interval starts soon")
+					continue
+				}
+				// Set expiry to prevent tickets from being mined in the next
+				// sdiff interval.  When the next block begins the new interval,
+				// the ticket is being purchased for the next interval; therefore
+				// increment expiry by a full sdiff window size to prevent it
+				// being mined in the interval after the next.
+				expiry = nextIntervalStart
+				if height+1 == nextIntervalStart {
+					expiry += intervalSize
+				}
+			}
+
+			cancelCtx, cancel := context.WithCancel(ctx)
+			cancels = append(cancels, cancel)
+			go func() {
+				err := vsp.buyTicket(cancelCtx, tipHeader, expiry, balanceToMaintain)
+				if err != nil {
+					switch {
+					// silence these errors
+					case errors.Is(err, errors.InsufficientBalance):
+					case errors.Is(err, context.Canceled):
+					case errors.Is(err, context.DeadlineExceeded):
+					default:
+						log.Errorf("Ticket purchasing failed: %v", err)
+					}
+					if errors.Is(err, errors.Passphrase) {
+						fatalMu.Lock()
+						fatal = err
+						fatalMu.Unlock()
+						outerCancel()
+					}
+				}
+			}()
+		}
+	}
+}
+
+// buyTicket calculate the number of tickets before purchase
+func (vsp *VSP) buyTicket(ctx context.Context, tip *wire.BlockHeader, expiry int32, maintain int64) error {
+	ctx, task := trace.NewTask(ctx, "ticketbuyer.buy")
+	defer task.End()
+
+	wal := vsp.w
+
+	networkBackend, err := wal.Internal().NetworkBackend()
+	if err != nil {
+		return err
+	}
+
+	// Determine how many tickets to buy
+	bal, err := wal.GetAccountBalance(int32(vsp.purchaseAccount))
+	if err != nil {
+		return err
+	}
+
+	spendable := bal.Spendable
+	if spendable < maintain {
+		log.Debugf("Skipping purchase: low available balance")
+		return nil
+	}
+
+	spendable -= maintain
+	sdiff, err := wal.Internal().NextStakeDifficultyAfterHeader(ctx, tip)
+	if err != nil {
+		return err
+	}
+
+	buy := int(dcrutil.Amount(spendable) / sdiff)
+	if buy == 0 {
+		log.Debugf("Skipping purchase: low available balance")
+		return nil
+	}
+
+	max := int(wal.Internal().ChainParams().MaxFreshStakePerBlock)
+	if buy > max {
+		buy = max
+	}
+
+	//TODO: limit should be set dynamically
+	limit := 1
+	if limit > 0 && buy > limit {
+		buy = limit
+	}
+
+	request := &w.PurchaseTicketsRequest{
+		Count:                buy,
+		SourceAccount:        vsp.purchaseAccount,
+		Expiry:               expiry,
+		MinConf:              DefaultRequiredConfirmations,
+		VSPFeeProcess:        vsp.PoolFee,
+		VSPFeePaymentProcess: vsp.ProcessFee,
+	}
+
+	tix, err := vsp.w.Internal().PurchaseTickets(ctx, networkBackend, request)
```

Comment:
It's worth noting that this wallet method will first mix funds before purchasing a ticket if cspp parameters are provided in `request`. If the user has enabled mixing, it would be vital to add the cssp params here so wallet can do some mixing before purchasing tickets. This is necessary even if purchasing from the mixed account because unless the mixed account has utxos with the exact ticket amount, one or more utxos will be used to create a new utxo with the exact ticket amount before the ticket will be purchased. This means that 1 or more mixed outputs may be combined into one transaction, and used to create an output of the same value as the ticket price, which will then be used to purchase the ticket. This risks undoing the results of the mixing in the first place.

Reply:
Perhaps this can be added to #225 after this goes in because #225 already makes a note to do this for regular ticket purchases, i.e. not automated ticket purchases.